### PR TITLE
fix issue 4303 : reshaping part

### LINKED ISSUE #4303: Bilinear_Upsampling requires to specify batch_size and num_input_channels
Referring to this line:
https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/abstract_conv.py#L570

this piece of code doesn't work with `batch_size` and/or `num_input_channels` set to `None`
it requires to specify the dimensions as scalar values

### PR DESCRIPTION
**Pull request by taesupkim**
Error fix #4303  with batch_size==None and num_input_channels==None during reshaping

### CODE REVIEW

File: theano/tensor/nnet/abstract_conv.py [hunk 1]
```diff
@@ -567,7 +567,7 @@ def bilinear_upsampling(input,
                                                subsample=(ratio, ratio),
                                                filter_flip=True)
 
-    return upsampled_mat.reshape((batch_size, num_input_channels,
+    return upsampled_mat.reshape((input.shape[0], input.shape[1],
```
Comment (nouiz) [inline theano/tensor/nnet/abstract_conv.py]: This need a test.

**Comment by taesupkim** [comment]
@fvisin Could you let me know ho the test this? Thanks.

**Comment by fvisin** [comment]
@taesupkim take a look at the [existing test for the bilinear upsampling 1d](https://github.com/taesupkim/Theano/blob/issue_4303/theano/tensor/nnet/tests/test_abstract_conv.py#L672).

Adding a new test that checks that you can call `bilinear_upsampling` with `batch_size = None` and `num_input_channels = None` should be enough. There is no need to test for multiple `ratio` in this case, just pick one.

**Comment by lamblin** [comment]
@taesupkim You can run the tests locally using `theano-nose` or `nosetests`.
For instance, to run only the tests for bilinear upsampling, run `theano-nose theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling`. To run one test method in particular, you can run `theano-nose theano/tensor/nnet/tests/test_abstract_conv.py:TestBilinearUpsampling.test_bilinear_upsampling_1D`

I've killed the tests on Travis so that it does not get overloaded.

**Comment by taesupkim** [comment]
@lamblin @fvisin Thanks for you comments. I had done testing it on local as you told and it is worked properly.

**Comment by lamblin** [comment]
OK, so if you have a new test case that works locally, then you can add it to the PR.

**Comment by taesupkim** [comment]
@fvisin I add test lines for this fixed part and it is also tested locally. Could you please check this out? I'll squash the commits after it work well. Thanks

File: theano/tensor/nnet/tests/test_abstract_conv.py [hunk 2]
```diff
@@ -687,6 +687,24 @@ def test_bilinear_upsampling_1D(self):
             up_mat_2d = self.get_upsampled_twobytwo_mat(input_x, ratio)
             utt.assert_allclose(f(), up_mat_2d, rtol=1e-06)
 
+    def test_bilinear_upsampling_reshaping(self):
+        """Test bilinear upsampling without giving shape information
```
Comment (fvisin) [inline theano/tensor/nnet/tests/test_abstract_conv.py]: Can you please change the docstring into a comment (i.e. use `#` to comment instead of `"""`)?
We do this only on the unit tests, because that allows the name of the test (e.g. `test_bilinear_upsampling_reshaping`) and its full path to get printed in case of failure.

File: theano/tensor/nnet/tests/test_abstract_conv.py [hunk 3]
```diff
@@ -687,6 +687,24 @@ def test_bilinear_upsampling_1D(self):
...
+        This method tests the bilinear_upsampling method
+        without giving batch_size and num_input_channels
+
+        """
+        # upsampling for a ratio of two
+        input_x = np.array([[[[1, 2], [3, 4]]]], dtype=theano.config.floatX)
+
+        for ratio in [2, 3]:
+            for use_1D_kernel in [True, False]:
+                bilin_mat = bilinear_upsampling(input=input_x, ratio=ratio,
```
Comment (fvisin) [inline theano/tensor/nnet/tests/test_abstract_conv.py]: Please set both `batch_size` and `num_input_channels` to `None`, so that in case the default values are changed in the future the test remains valid.

**Comment by fvisin** [comment]
LGTM. When my last comments are addressed we can merge if Travis passes.
Thank you @taesupkim.

**Comment by lamblin** [comment]
@taesupkim Can you please make the changes requested, and update the PR?

**Comment by lamblin** [comment]
Thanks!
